[RFR]Added test_reload_button_provider.py to test functionality of refresh button

jdemon519 opened the pull request:
{{pytest: cfme/tests/containers/test_reload_button_provider.py -v --use-provider cm-env2 }}

The test verifies data integrity of the fields in the Relationships table (while on provider summary page) after the reload button is clicked. 
Update: added new functionality to verify data integrity for images

dajoRH commented:
Lint report for commit 73db986b14678a5be2adb09e977b0856cb891c75:

Lint OK! :cake: :punch: :cookie: .


**No commit flaws detected**.
<mergeable/>

*CFME QE Bot*

dajoRH commented:
Lint report for commit d42d682b19bbf6686abc823609a4ae23c65a1415:

Lint OK! :cake: :punch: :cookie: .


**No commit flaws detected**.
<mergeable/>

*CFME QE Bot*

gshefer commented on cfme/tests/containers/test_reload_button_provider.py:
```diff
@@ -0,0 +1,45 @@
+    tb.select('Reload Current Display')
+    provider.validate_stats(ui=True)
+
+    num_img_from_registry = len(provider.mgmt.o_api.get('image')[1]['items'])
```
I'm not sure if it gives the same results but you have provider.mgmt.list_image_registry (see [this wrapanapi file](https://github.com/ManageIQ/wrapanapi/blob/master/mgmtsystem/kubernetes.py) line 157).

jdemon519 replied:
Just checked, len(provider.mgmt.list_image_registry) returns 1, which is the number of registries. I used len(provider.mgmt.o_api.get('image')[1]['items']) to get the number of images in that registry, which in this case is 85.

gshefer commented on cfme/tests/containers/test_reload_button_provider.py:
```diff
@@ -0,0 +1,45 @@
+    num_img_from_openshift = len(provider.mgmt.list_image())
+
+    num_img_api_list = num_img_from_registry + num_img_from_openshift
+    num_img_ui = provider.num_image()
```
I'm not sure if it gives the same results but you have provider.num_image_ui (see [ContainersProvider](https://github.com/ManageIQ/integration_tests/blob/master/cfme/containers/provider/__init__.py) line 247).

jdemon519 replied:
Thought about using it too, but it didn't work, I am getting this error "'NoneType' object is not callable", while provider.num_image() returns the correct number of images.

mshriver replied:
This is providing results from the database, not the UI, which makes the test misleading. Let's debug why that exception is happening when calling num_image_ui.

gshefer commented on cfme/containers/provider/__init__.py:
```diff
@@ -110,8 +110,8 @@ class ContainersProvider(BaseProvider, Pretty):
-        'num_container',
-        'num_image']
+        'num_image_registry',
+        'num_container']
```
Why not just add num_image_registry instead of replacing num_image?

jdemon519 replied:
I thought I'd override that function, by taking out num_image for now. Reason being, num_image works for 5.6 but not for 5.7, and I haven't made any changes in the API related to images. Should I? I added num_image_registry which was not there before. I have added the Template object to the API, let me know what you think of this

https://github.com/ManageIQ/wrapanapi/pull/96

mshriver replied:
This should be version picked here if its num_image in 5.6 and num_image_registry in 5.7

mshriver commented:
@jdemon519 PRT results are a false positive for run 9372, tests skipped because cm-env2 failed to setup.

mshriver commented on cfme/tests/containers/test_reload_button_provider.py:
```diff
@@ -0,0 +1,45 @@
+import pytest
+
+from utils.appliance.implementations.ui import navigate_to
```
We encourage contributors to organize imports alphabetizing/grouping at least the first module in the path.

mshriver commented on cfme/tests/containers/test_reload_button_provider.py:
```diff
@@ -0,0 +1,45 @@
+from utils.appliance.implementations.ui import navigate_to
+from cfme.containers.provider import ContainersProvider
+from utils import testgen, version
+from utils.version import current_version
```
You import version, and version.current_version separately, and then call current_version both using the imported module and the function. Pick one way or the other, but since its the only function from version that you're using, I'd recommend keeping the function import and dropping the version import from line 5.

mshriver commented on cfme/tests/containers/test_reload_button_provider.py:
```diff
@@ -0,0 +1,45 @@
+
+
+def test_reload_button_provider(provider):
+    """ This test verifies the data integrity of the fields in
```
:+1: great test description.

mshriver commented on cfme/tests/containers/test_reload_button_provider.py:
```diff
@@ -0,0 +1,45 @@
+    """
+
+    navigate_to(ContainersProvider, 'All')
+    provider.load_details()
```
You don't need lines 31 and 32, I recommend calling `navigate_to(provider, 'Details')`

mshriver commented on cfme/tests/containers/test_reload_button_provider.py:
```diff
@@ -0,0 +1,45 @@
+
+    num_img_from_registry = len(provider.mgmt.o_api.get('image')[1]['items'])
+    num_img_from_openshift = len(provider.mgmt.list_image())
+
```
consider logging these values for easier test result debugging

mshriver commented on cfme/tests/containers/test_reload_button_provider.py:
```diff
@@ -0,0 +1,45 @@
+    if version.current_version() < "5.7":
+        assert num_img_ui == num_img_from_openshift
+    else:
+        assert num_img_api_list == num_img_ui
```
Consider a change in syntax here, assuming a switch to `num_image_ui()` as well.

```python
assert provider.num_image_ui() == version.pick({version.LOWEST: num_img_from_openshift, 
                                                                         '5.7': num_img_api_list})
```

jdemon519 commented:
@mshriver Thanks for the feedback, I have switched to provider.summary.relationships.container_images.value to address the UI issue, fixed the imports and added templates to STATS_TO_MATCH list for verification. The PRT is fixed so that should address the test runs issue.

RonnyPfannschmidt commented on cfme/containers/provider/openshift.py:
```diff
@@ -14,7 +14,7 @@ def __init__(self, name, value, field_type=None, href=None):
 @ContainersProvider.add_provider_type
 class OpenshiftProvider(ContainersProvider):
-    STATS_TO_MATCH = ContainersProvider.STATS_TO_MATCH + ['num_route']
+    STATS_TO_MATCH = ContainersProvider.STATS_TO_MATCH + ['num_route'] + ['num_template']
```
please merge the 2 single element lists into one 2 element list

jdemon519 replied:
@RonnyPfannschmidt Done

psav commented on cfme/containers/provider/openshift.py:
```diff
@@ -14,7 +14,8 @@ def __init__(self, name, value, field_type=None, href=None):
 class OpenshiftProvider(ContainersProvider):
-    STATS_TO_MATCH = ContainersProvider.STATS_TO_MATCH + ['num_route']
+    num_route_template = ['num_route'] + ['num_template']
+    STATS_TO_MATCH = ContainersProvider.STATS_TO_MATCH + num_route_template
```
Happy to leave it but it doesn't seem necessary to create a new variable for num_route_template here
